lnwallet/transactions_test.go

```diff
@@ -649,218 +446,189 @@ var testCases = []testCase{
-				spew.Sdump(commitTx))
-			continue
-		}
+	err = localChannel.ReceiveNewCommitment(remoteSig, remoteHtlcSigs)
+	require.NoError(t, err)
 
-		// Generate second-level HTLC transactions for HTLCs in
-		// commitment tx.
-		htlcResolutions, err := extractHtlcResolutions(
-			chainfee.SatPerKWeight(test.commitment.FeePerKw), true,
-			signer, htlcs, keys, &channel.channelState.LocalChanCfg,
-			&channel.channelState.RemoteChanCfg, commitTx.TxHash(),
-			channel.channelState.ChanType,
-		)
-		if err != nil {
-			t.Errorf("Case %d: Failed to extract HTLC resolutions: %v", i, err)
-			continue
-		}
+	_, _, err = localChannel.RevokeCurrentCommitment()
+	require.NoError(t, err)
 
-		resolutionIdx := 0
-		for j, htlcDesc := range test.htlcDescs {
-			// TODO: Check HTLC success transactions; currently not implemented.
-			// resolutionIdx can be replaced by j when this is handled.
-			if htlcs[j].Incoming {
-				continue
-			}
+	// Now the local node force closes the channel so that we can inspect
+	// its state.
+	forceCloseSum, err := localChannel.ForceClose()
+	require.NoError(t, err)
 
-			expectedTx, err := txFromHex(htlcDesc.resolutionTxHex)
-			if err != nil {
-				t.Fatalf("Failed to parse serialized tx: %v", err)
-			}
+	// Assert that the commitment transaction itself is as expected.
+	var txBytes bytes.Buffer
+	require.NoError(t, forceCloseSum.CloseTx.Serialize(&txBytes))
 
-			htlcResolution := htlcResolutions.OutgoingHTLCs[resolutionIdx]
-			resolutionIdx++
+	fmt.Printf("commit tx: %x\n", txBytes.Bytes())
 
-			actualTx := htlcResolution.SignedTimeoutTx
-			if actualTx == nil {
-				t.Errorf("Case %d: Failed to generate second level tx: "+
-					"output %d, %v", i, j,
-					htlcResolutions.OutgoingHTLCs[j])
-				continue
-			}
+	require.Equal(t, test.expectedCommitmentTxHex, hex.EncodeToString(txBytes.Bytes()))
 
-			// Check that second-level HTLC transaction was created correctly.
-			if actualTx.WitnessHash() != *expectedTx.WitnessHash() {
-				t.Errorf("Case %d: Generated unexpected second level tx: "+
-					"output %d, expected %s, got %s", i, j,
-					expectedTx.WitnessHash(), actualTx.WitnessHash())
-				continue
-			}
-		}
+	// Obtain the second level transactions that the local node's channel
+	// state machine has produced. Store them in a map indexed by commit tx
+	// output index. Also complete the second level transaction with the
+	// preimage. This is normally done later in the contract resolver.
+	secondLevelTxes := map[uint32]*wire.MsgTx{}
+	for _, r := range forceCloseSum.HtlcResolutions.IncomingHTLCs {
+		successTx := r.SignedSuccessTx
+		witnessScript := successTx.TxIn[0].Witness[4]
+		var hash160 [20]byte
+		copy(hash160[:], witnessScript[69:69+20])
+		preimage := hash160map[hash160]
+		successTx.TxIn[0].Witness[3] = preimage[:]
+		secondLevelTxes[r.HtlcPoint().Index] = successTx
+	}
+	for _, r := range forceCloseSum.HtlcResolutions.OutgoingHTLCs {
+		secondLevelTxes[r.HtlcPoint().Index] = r.SignedTimeoutTx
+	}
```

Comment:
add an assertion on the size of `secondLevelTxes` to ensure nothing was overwritten?

Reply:
Added `storeTx` closure with assertion.